app/vmalert/unittest.go

```diff
@@ -0,0 +1,780 @@
+		resp, err := http.Get(testHealthHTTPPath)
+		if err != nil {
+			return false
+		}
+		_ = resp.Body.Close()
+		return resp.StatusCode == 200
+	}
+	if err := waitFor(testStorageInitTimeout, readyStorageCheckFunc); err != nil {
+		logger.Fatalf("http server can't start for %s seconds, err %s", testStorageInitTimeout, err)
+	}
+}
+
+func tearDown() {
+	if err := httpserver.Stop(*httpListenAddr); err != nil {
+		logger.Errorf("cannot stop the webservice: %s", err)
+	}
+	vmstorage.Stop()
+	metrics.UnregisterAllMetrics()
+	fs.MustRemoveAll(storagePath)
+}
+
+// resolveAndGlobFilepaths joins all relative paths in a configuration
+// with a given base directory and replaces all globs with matching files.
+func resolveAndGlobFilepaths(baseDir string, utf *unitTestFile) error {
+	for i, rf := range utf.RuleFiles {
+		if rf != "" && !filepath.IsAbs(rf) {
+			utf.RuleFiles[i] = filepath.Join(baseDir, rf)
+		}
+	}
+
+	var globbedFiles []string
+	for _, rf := range utf.RuleFiles {
+		m, err := filepath.Glob(rf)
+		if err != nil {
+			return err
+		}
+		if len(m) == 0 {
+			fmt.Fprintln(os.Stderr, "  WARNING: no file match pattern", rf)
+		}
+		globbedFiles = append(globbedFiles, m...)
+	}
+	utf.RuleFiles = globbedFiles
+	return nil
+}
+
+func ruleUnitTest(filename string) []error {
+	fmt.Println("\nUnit Testing: ", filename)
+	b, err := os.ReadFile(filename)
+	if err != nil {
+		return []error{fmt.Errorf("failed to read unit test file %s: %v", filename, err)}
+	}
+
+	var unitTestInp unitTestFile
+	if err := yaml.UnmarshalStrict(b, &unitTestInp); err != nil {
+		return []error{fmt.Errorf("failed to unmarshal unit test file %s: %v", filename, err)}
+	}
+	if err := resolveAndGlobFilepaths(filepath.Dir(filename), &unitTestInp); err != nil {
+		return []error{fmt.Errorf("failed to unmarshal unit test file %s: %v", filename, err)}
+	}
+
+	if unitTestInp.EvaluationInterval.Duration() == 0 {
+		unitTestInp.EvaluationInterval = &promutils.Duration{D: 1 * time.Minute}
+	}
+
+	groupOrderMap := make(map[string]int)
+	for i, gn := range unitTestInp.GroupEvalOrder {
+		if _, ok := groupOrderMap[gn]; ok {
+			return []error{fmt.Errorf("group name repeated in evaluation order: %s", gn)}
+		}
+		groupOrderMap[gn] = i
+	}
+
+	var errs []error
+	for _, t := range unitTestInp.Tests {
+		errs = append(errs, t.test(unitTestInp.EvaluationInterval.Duration(), groupOrderMap, unitTestInp.RuleFiles...)...)
+	}
+
+	if len(errs) > 0 {
+		return errs
+	}
+	return nil
+}
+
+func httpWrite(address string, r io.Reader) {
+	resp, err := http.Post(address, "", r)
+	if err != nil || resp.StatusCode != 204 {
+		logger.Errorf("failed to send to storage: %v", err)
+	}
+	resp.Body.Close()
+}
+
+func (tg *testGroup) test(evalInterval time.Duration, groupOrderMap map[string]int, ruleFiles ...string) []error {
+	// set up vmstorage and http server
+	setUp()
+	// tear down vmstorage and clean the data dir
+	defer tearDown()
+	r := testutil.WriteRequest{}
+	for _, data := range tg.InputSeries {
```

Comment:
I think many details here, such as config struct, parsing functions, data generation functions, etc. 
can be moved to unittest package.
The /vmalert/unittest.go should only operate and control the flow of the function execution. This would make it simpler to read and test.

Reply:
I tried to extract most of them to /unittest in the new commit, please check)
But checking alerting test case is a little bit tricky, it has many relation with Group evaluation in the main package, move it under /unittest doesn't look good.